src/sentry/monitors/system_incidents.py

```diff
@@ -166,6 +177,258 @@ def get_clock_tick_volume_metric(tick: datetime) -> float | None:
+    """
+
+    INCIDENT_RECOVERING = enum.auto()
+    """
+    An incident has begun to recover. After this transition we will either
+    re-enter the incident va INCIDENT_STARTED or fully recover via
+    INCIDENT_RECOVERED.
+    """
+
+    INCIDENT_RECOVERED = enum.auto()
+    """
+    An incident has recovered back to normal.
+    """
+
+
+@dataclass
+class DecisionResult:
+    decision: TickAnomalyDecision
+    """
+    The recorded decision made for the clock tick
+    """
+
+    transition: AnomalyTransition | None = None
+    """
+    Reflects the transition status when making a tick decision results in a
+    state transition. None if the decision has not changed.
+    """
+
+
+def update_clock_tick_decisions(tick: datetime) -> DecisionResult:
+    """
+    Given a clock tick timestamp determine based on the historic tick volume
+    metrics, and historic tick anomaly decisions, a DecisionResult.
+
+    Once there is an INCIDENT future ticks will continue to be marked as
+    INCIDENT until we begin detecting metrics below the abnormal metric
+    threshold, at which point the system will recover after
+    MONITOR_TICK_DECISION_WINDOW ticks below the abnormal threshold.
+
+    This function will update previous decisions for earlier ticks detected as
+    ABNORMAL to either NORMAL or INCIDENT.
+    """
+    # Alias TickAnomalyDecision to improve code readability
+    Decision = TickAnomalyDecision
+
+    if not options.get("crons.tick_volume_anomaly_detection"):
+        return DecisionResult(Decision.NORMAL)
+
+    redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+
+    # The clock has just ticked to the next minute. Look at the previous tick
+    # and decision metrics.
+    past_ts = tick - timedelta(minutes=1)
+
+    past_window_ts_keys = [
+        _make_reference_ts(past_ts - timedelta(minutes=delta))
+        for delta in range(0, MONITOR_TICK_DECISION_WINDOW)
+    ]
+
+    # Fetch histories for metrics and the last decision together. Window
+    # timestamps are reversed so the oldest metric is last.
+    redis_keys = chain(
+        (MONITOR_TICK_METRIC.format(ts=ts) for ts in reversed(past_window_ts_keys)),
+        (MONITOR_TICK_DECISION.format(ts=ts) for ts in [past_window_ts_keys[0]]),
+    )
+
+    histories = redis_client.mget(redis_keys)
+
+    # Tick metrics are the first MONITOR_TICK_DECISION_WINDOW values
+    tick_metrics = [float(value) if value else None for value in histories[:-1]]
+
+    # The last decision is the last value fetched
+    if histories[-1] is not None:
+        last_decision = Decision.from_str(histories[-1])
+    else:
+        # By default the previous decision is used. If there was no previous
+        # decision we can only assume things are operating normally
+        last_decision = Decision.NORMAL
+
+    # Pre-compute decisions for the history of metrics
+    metric_decisions = [_metric_decision(metric) for metric in tick_metrics]
```

Comment:
I feel like `metric_decisions` and `last_decision` are confusing. `last_decision` is the actual decision of the last time that this process ran, whereas `metric_decisions` is more like the decision state of each metric, not the final decision. I'm not totally sure on naming, but it took me a lot of reading to figure out what was happening and the difference

Reply:
I think this should be better now